include/swift/AST/Types.h

```diff
@@ -1450,11 +1450,26 @@ class ParameterTypeFlags {
                                        : value - ParameterTypeFlags::Shared);
   }
 
+  ParameterTypeFlags withAutoClosure(bool isAutoClosure) const {
+    return ParameterTypeFlags(isAutoClosure
+                                  ? value | ParameterTypeFlags::AutoClosure
+                                  : value - ParameterTypeFlags::AutoClosure);
+  }
+
   bool operator ==(const ParameterTypeFlags &other) const {
     return value.toRaw() == other.value.toRaw();
   }
 
   uint8_t toRaw() const { return value.toRaw(); }
+
+  static ParameterTypeFlags fromRaw(uint8_t value) {
+    return ParameterTypeFlags()
+        .withVariadic(value & Variadic)
+        .withAutoClosure(value & AutoClosure)
+        .withEscaping(value & Escaping)
```

Comment:
I don't feel like autoclosure or escaping belong in the parameter flags, but that concern is more-or-less independent of your improvements to function type metadata.

Reply:
@DougGregor I think I can mask them out from the metadata with ease, do you think the approach here is viable overall?